**Tugas 1:**
Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv. Evaluasi model ini dan laporkan hasilnya.

In [9]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [10]:
# Membaca data dari file CSV
data = pd.read_csv('data/voice.csv')
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [11]:
# Memisahkan fitur (X) dan label (y)
X = data.drop('label', axis=1)
y = data['label']

# Memisahkan data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [12]:
# Membuat objek SVM
svm_classifier = SVC(kernel='linear', C=1, gamma='scale')

# Melatih model SVM pada data pelatihan
svm_classifier.fit(X_train, y_train)

SVC(C=1, kernel='linear')

In [13]:
# Melakukan prediksi pada data pengujian
y_pred = svm_classifier.predict(X_test)

# Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy:.2f}')

# Menampilkan laporan klasifikasi
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi:')
print(classification_rep)

# Menampilkan matriks kebingungan
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan:')
print(confusion_mat)

Akurasi: 0.94
Laporan Klasifikasi:
              precision    recall  f1-score   support

      female       0.98      0.89      0.93       297
        male       0.91      0.99      0.95       337

    accuracy                           0.94       634
   macro avg       0.95      0.94      0.94       634
weighted avg       0.94      0.94      0.94       634

Matriks Kebingungan:
[[265  32]
 [  5 332]]


**Tugas 2:**
Buatlah model klasifikasi Multinomial Naive Bayes dengan ketentuan:
1. Menggunakan data spam.csv
2. Fitur CountVectorizer dengan mengaktifkan stop_words
3. Evaluasi hasilnya dan laporkan hasil evaluasi model ini.

In [14]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Membaca data dari file CSV
data = pd.read_csv('data/spam.csv', encoding='latin-1')

data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [15]:
# Menghapus kolom "Unnamed: 2", "Unnamed: 3", dan "Unnamed: 4"
data = data.drop(columns=["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"])

# Mengganti nama kolom "v1" dan "v2" dengan nama yang sesuai
data = data.rename(columns={"v1": "labels", "v2": "sms"})

data.head()

labels                                                sms
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

In [16]:
# Memisahkan kolom 'sms' sebagai teks pesan dan 'labels' sebagai label
X = data['sms']
y = data['labels']

# Memisahkan data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [17]:
# Membuat objek CountVectorizer dengan mengaktifkan stop words
vectorizer = CountVectorizer(stop_words='english')

# Melakukan transformasi teks pesan menjadi vektor fitur
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

In [18]:
# Membuat objek Multinomial Naive Bayes
naive_bayes_classifier = MultinomialNB()

# Melatih model Multinomial Naive Bayes pada data pelatihan
naive_bayes_classifier.fit(X_train, y_train)

MultinomialNB()

In [19]:
# Melakukan prediksi pada data pengujian
y_pred = naive_bayes_classifier.predict(X_test)

# Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy:.2f}')

# Menampilkan laporan klasifikasi
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi:')
print(classification_rep)

# Menampilkan matriks kebingungan
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan:')
print(confusion_mat)

Akurasi: 0.98
Laporan Klasifikasi:
              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115

Matriks Kebingungan:
[[959   6]
 [ 12 138]]


**Evaluasi Hasil**:

- **Akurasi**: 0.98 (98% akurasi).

- **Laporan Klasifikasi**:
  - Precision: 0.99 untuk "ham" (tidak spam) dan 0.96 untuk "spam."
  - Recall: 0.99 untuk "ham" dan 0.92 untuk "spam."
  - F1-score: 0.99 untuk "ham" dan 0.94 untuk "spam."

- **Matriks Kebingungan (Confusion Matrix)**:
  - 12 false negatives (spam salah diklasifikasikan sebagai ham).
  - 6 false positives (ham salah diklasifikasikan sebagai spam).

Secara keseluruhan, model ini sangat baik dalam mengidentifikasi "ham," tetapi ada beberapa kesalahan dalam mengidentifikasi "spam." Meskipun ada false negatives, akurasi sekitar 0.98 menunjukkan bahwa model ini masih efektif secara keseluruhan dalam membedakan email spam dan non-spam.

**Tugas 3:**
Buatlah model klasifikasi Multinomial Naive Bayes dengan ketentuan:
1. Menggunakan data spam.csv
2. Fitur TF-IDF dengan mengaktifkan stop_words
3. Evaluasi hasilnya dan bandingkan hasil evaluasi model ini dengan model yang dibuat pada Tugas 2.
4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [20]:
# Memisahkan kolom 'sms' sebagai teks pesan dan 'label' sebagai label
X = data['sms']
y = data['labels']

# Memisahkan data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Membuat objek TfidfVectorizer dengan mengaktifkan stop words
vectorizer = TfidfVectorizer(stop_words='english')

# Melakukan transformasi teks pesan menjadi vektor fitur TF-IDF
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

In [22]:
# Membuat objek Multinomial Naive Bayes
naive_bayes_classifier = MultinomialNB()

# Melatih model Multinomial Naive Bayes pada data pelatihan dengan fitur TF-IDF
naive_bayes_classifier.fit(X_train_tfidf, y_train)

MultinomialNB()

In [23]:
# Melakukan prediksi pada data pengujian
y_pred = naive_bayes_classifier.predict(X_test_tfidf)

# Menghitung akurasi dengan fitur TF-IDF
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi dengan TF-IDF: {accuracy:.2f}')

# Menampilkan laporan klasifikasi dengan fitur TF-IDF
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi dengan TF-IDF:')
print(classification_rep)

# Menampilkan matriks kebingungan dengan fitur TF-IDF
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan dengan TF-IDF:')
print(confusion_mat)

Akurasi dengan TF-IDF: 0.97
Laporan Klasifikasi dengan TF-IDF:
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115

Matriks Kebingungan dengan TF-IDF:
[[965   0]
 [ 37 113]]


**Evaluasi Hasil:**

- **Akurasi dengan TF-IDF**: 0.97 (97% akurasi).

- **Laporan Klasifikasi dengan TF-IDF**:
  - Precision: 0.96 untuk "ham" (tidak spam) dan 1.00 untuk "spam."
  - Recall: 1.00 untuk "ham" dan 0.75 untuk "spam."
  - F1-score: 0.98 untuk "ham" dan 0.86 untuk "spam."

- **Matriks Kebingungan (Confusion Matrix) dengan TF-IDF**:
  - 37 false negatives (spam salah diklasifikasikan sebagai ham) dan 0 false positives (ham tidak salah diklasifikasikan sebagai spam).

Model dengan fitur TF-IDF memiliki recall yang lebih rendah untuk kategori "spam" dibandingkan dengan model yang menggunakan CountVectorizer. Meskipun demikian, model TF-IDF mencapai akurasi sekitar 0.97, yang masih merupakan kinerja yang baik, tetapi model dengan CountVectorizer lebih baik dalam mengidentifikasi email spam secara keseluruhan.

**Kesimpulan:**

- **Akurasi**:
  - Model dengan **CountVectorizer** memiliki akurasi sekitar **0.98**.
  - Model dengan **TF-IDF** memiliki akurasi sekitar **0.97**.

- **Laporan Klasifikasi**:
  - Model dengan **CountVectorizer** memiliki **f1-score** yang baik untuk kategori "spam" (0.94), tetapi memiliki **recall** yang lebih rendah untuk kategori "spam" (0.92).
  - Model dengan **TF-IDF** memiliki **f1-score** yang lebih rendah untuk kategori "spam" (0.86) dan recall yang lebih rendah (0.75) untuk kategori "spam".

- **Matriks Kebingungan (Confusion Matrix)**:
  - Model dengan **CountVectorizer** memiliki 12 false negatives (spam yang salah diklasifikasikan sebagai ham), yang lebih rendah dibandingkan dengan model TF-IDF.
  - Model dengan **TF-IDF** memiliki 37 false negatives, yang lebih tinggi dibandingkan dengan model CountVectorizer.

Berdasarkan hasil ini, **model dengan CountVectorizer** tampaknya memberikan kinerja yang lebih baik dalam hal **recall** untuk kategori "spam" (menemukan lebih banyak email spam secara akurat) dan memiliki **f1-score** yang lebih tinggi untuk kategori "spam". Oleh karena itu, dalam kasus data "spam.csv," **fitur CountVectorizer** mungkin lebih baik daripada TF-IDF untuk tugas klasifikasi spam.